src/main/java/com/google/devtools/common/options/Converters.java

```diff
@@ -625,4 +635,73 @@ public String getTypeDescription() {
       return "Converts to a CacheBuilderSpec, or null if the input is empty";
     }
   }
+
+  public static class ExpressionConverter implements Converter<Expression> {
+    @Override
+    public Expression convert(String input) throws OptionsParsingException {
+      if (input.isEmpty()) {
+        return Expression.empty();
+      }
+      if (isLegacySyntax(input)) {
+        input = convertLegacySyntaxToBooleanExpression(input);
+      }
+      try {
+        Expression result = Expression.parse(ParserInput.fromLines(input));
+        checkForForbiddenOperators(result);
+        return result;
+      } catch (SyntaxError.Exception e) {
+        throw new OptionsParsingException("Failed to parse expression: " + e.getMessage() + " input: " + input, e);
+      }
+    }
+
+    @Override
+    public String getTypeDescription() {
+      return "Converts to an Expression.";
+    }
+
+    private boolean isLegacySyntax(String input) {
+      return input.contains(",") || input.contains("-");
+    }
+
+    /**
+     * Converts a string containing tags separated by comma to a boolean expression.
+     */
+    private String convertLegacySyntaxToBooleanExpression(String input) throws OptionsParsingException {
+      CommaSeparatedOptionListConverter legacyConverter = new CommaSeparatedOptionListConverter();
+      List<String> tags = legacyConverter.convert(input);
+      return (" " + String.join(" or ", tags)).replace(" -", " not ").substring(1);
+    }
+
+    /**
+     * Throws an OptionsParsingException when the given expression does not follow the grammar:
+     * expr = IDENT | expr 'or' expr | expr 'and' expr | 'not' expr | '(' expr ')'
+     */
+    private void checkForForbiddenOperators(Expression expression) throws OptionsParsingException {
+      switch (expression.kind()) {
+        case IDENTIFIER:
+          break;
+        case BINARY_OPERATOR:
+          BinaryOperatorExpression boe = (BinaryOperatorExpression) expression;
+          if (boe.getOperator() == OR || boe.getOperator() == AND) {
+              checkForForbiddenOperators(boe.getX());
+              checkForForbiddenOperators(boe.getY());
+          } else {
+            throw new OptionsParsingException(String.format("Syntax Error: Got %s but expected 'and' or 'or'",
```

Comment:
Use lowercase in error messages. e.g.

"invalid Boolean operator: %s (want 'and' or 'or')"